### ISSUE TITLE
simple automatic parallelization through distinct sub-graph detection

### ISSUE THREAD

**Issue by sgpthomas** [cucapra/futil#75]
This pass would detect enable sub-graphs that are specified to run in sequence, but can actually be run in parallel without changing the observable behavior of the program. For example, consider the following program:
```lisp
(seq (enable A B)
     (enable C D)
      ...)
```
Suppose that there are no edges between the sub-graph induced by `(enable A B)` and the sub-graph induced by `(enable C D)`. If this is the case, then there is no way for these subgraphs to depend on each other and then can be run in parallel. We can transform this program into:
```lisp
(seq (enable A B C D)
      ...)
```
There are certainly more complicated cases where you might be able to parallelize the control, but for this pass let's focus on this one simple case.

**Comment by tissue3** [comment]
Shall we add some data dependency analysis, or is it enough to check sequence of enable have common components?

**Comment by sgpthomas** [comment]
I think some data dependency analysis is needed to make this sound. I don't think it's enough to just check common subcomponents. I think you also have to make sure there are no data edges between the sub-graphs.

**Comment by tissue3** [comment]
I will implement this pass now. I assume as long as there does not exist edge between sub-graphs, we can merge enables. Comment on this issue if I miss anything

**Comment by sgpthomas** [comment]
This is being addressed in #102.

**Comment by rachitnigam** [comment]
Since the rewrite using groups and go/done proposal (#139), the discussion above is stale.

New problems from the groups discussion:
- We somehow need to know that a structural component is shareable. For example, a combinational adder is shareable but an adder that counts the number of times its called is not.
- Perhaps we need a `reset` hole which allows the context to know that any internal state is going to be reset.

**Comment by rachitnigam** [comment]
Too stale to be actionable. We should attempt to revive this in a new issue.
